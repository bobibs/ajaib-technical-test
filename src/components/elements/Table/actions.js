import moment from 'moment';
import { capitalizeFirstLetter, textTruncate } from '../../../utils/format';

export const convertData = (data) => {
  return data.map((i) => [
    {
      text: textTruncate(i.username, 20),
      key: 'username',
    },
    { text: textTruncate(i.name, 20), key: 'name' },
    { text: textTruncate(i.email, 20), key: 'email' },
    { text: capitalizeFirstLetter(i.gender), key: 'gender' },
    {
      text: moment(i.registeredDate).format('DD/MM/YYYY'),
      key: 'registeredDate',
    },
  ]);
};

export const sortAsc = (key) => (a, b) => a[key] > b[key] ? 1 : -1;

export const sortDesc = (key) => (a, b) => a[key] < b[key] ? 1 : -1;

export const sortData = (data, sortBy, sortFormat) => {
  const filterData = data.results.map((i) => ({
    username: `${i.name.title.toLowerCase()}${i.name.first.toLowerCase()}`,
    name: `${i.name.first} ${i.name.last}`,
    email: i.email,
    gender: i.gender,
    registeredDate: i.registered.date,
  }));

  if (sortFormat === 'asc') {
    return filterData.sort(sortAsc(sortBy));
  } else if (sortFormat === 'desc') {
    return filterData.sort(sortDesc(sortBy));
  } else {
    return filterData;
  }
};
