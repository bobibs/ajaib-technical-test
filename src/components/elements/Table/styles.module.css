.root {
    width: 100%;
}

.table {
    display: grid;
    grid-template-columns: repeat(var(--arrayLength), 1fr);
}

.tableHead {
    display: flex;
    align-items: center;
    justify-content: space-between;
    padding: 0 1rem 0 1rem;
    height: 3rem;
    background: var(--primary-color);
    margin: 0 0 1rem 0;
    color: #ffffff;
}

.tableHeadAction {
    display: flex;
    flex-direction: column;
    align-items: center;
    justify-content: center;
}

.tableHeadIcon {
    height: 0.875rem;
    width: 0.875rem;
    cursor: pointer;
    opacity: 0.5;
}

.tableHeadIcon:hover,
.tableHeadIconActive {
    opacity: 1;
}

.ascIcon {
    position: relative;
    top: 0.2rem;
}

.descIcon {
    position: relative;
    bottom: 0.2rem;
}

.tableBody {
    display: grid;
    grid-template-columns: 1fr;
    gap: 1.25rem 0;
    padding: 0 0 0 1rem;
}

.tableBodyText {
    font-size: 0.875rem;
}