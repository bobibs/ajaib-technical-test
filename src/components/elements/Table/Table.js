import React, { useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import ImageBox from '../ImageBox';
import { convertData, sortAsc, sortDesc, sortData } from './actions';
import styles from './styles.module.css';

export default function Table(props) {
  const { column, dataSource } = props;
  const [sortBy, setSortBy] = useState('');
  const [sortFormat, setSortFormat] = useState('');

  const handleSortAsc = (key) => {
    if (sortFormat !== 'asc') {
      setSortBy(key);
      setSortFormat('asc');
    } else {
      setSortBy('');
      setSortFormat('');
    }
  };

  const handleSortDesc = (key) => {
    if (sortFormat !== 'desc') {
      setSortBy(key);
      setSortFormat('desc');
    } else {
      setSortBy('');
      setSortFormat('');
    }
  };

  const data = column.map((i) => ({
    ...i,
    data: convertData(sortData(dataSource, sortBy, sortFormat)).map(
      (j) => j.filter((k) => k.key === i.key)[0],
    ),
  }));

  return (
    <div className={styles.root}>
      <Tables
        data={data}
        sortBy={sortBy}
        sortByAsc={handleSortAsc}
        sortByDesc={handleSortDesc}
        sortFormat={sortFormat}
      />
    </div>
  );
}

Table.defaultProps = {
  column: [],
  dataSource: {},
};

Table.propTypes = {
  column: PropTypes.array,
  dataSource: PropTypes.object,
};

export const Tables = (props) => {
  const { data, sortBy, sortByAsc, sortByDesc, sortFormat } = props;

  useEffect(() => {
    document.documentElement.style.setProperty('--arrayLength', data.length);
  }, []);

  return (
    <div className={styles.table}>
      {data.map((i, idx) => (
        <div key={idx}>
          <div className={styles.tableHead}>
            <div>{i.title}</div>
            {i.ascDesc ? (
              <div className={styles.tableHeadAction}>
                <ImageBox
                  className={`${styles.tableHeadIcon} ${styles.ascIcon} ${
                    sortBy === i.key && sortFormat === 'asc'
                      ? styles.tableHeadIconActive
                      : ''
                  }`}
                  onClick={() => sortByAsc(i.key)}
                  src='caret-up'
                />
                <ImageBox
                  className={`${styles.tableHeadIcon} ${styles.descIcon} ${
                    sortBy === i.key && sortFormat === 'desc'
                      ? styles.tableHeadIconActive
                      : ''
                  }`}
                  onClick={() => sortByDesc(i.key)}
                  src='caret-down'
                />
              </div>
            ) : (
              ''
            )}
          </div>
          <div className={styles.tableBody}>
            {i.data.map((j, jdx) => (
              <div className={styles.tableBodyText} key={jdx}>
                <div>{j && j.text}</div>
              </div>
            ))}
          </div>
        </div>
      ))}
    </div>
  );
};

Tables.defaultProps = {
  data: [],
  sortBy: '',
  sortByAsc: () => {},
  sortByDesc: () => {},
  sortFormat: '',
};

Tables.propTypes = {
  data: PropTypes.array,
  sortBy: PropTypes.string,
  sortByAsc: PropTypes.func,
  sortByDesc: PropTypes.func,
  sortFormat: PropTypes.string,
};
