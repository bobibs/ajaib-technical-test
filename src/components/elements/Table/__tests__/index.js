import React, { useEffect, useState } from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import Table, { Tables } from '../Table';

jest.mock('react', () => ({
  ...jest.requireActual('react'),
  useState: jest.fn(),
  useEffect: jest.fn(),
}));

const setSortBy = jest.fn();

describe('src/components/elements/Table', () => {
  beforeEach(() => {
    useState.mockImplementation((v) => [v, setSortBy]);
    useEffect.mockImplementation((fn) => fn());
  });

  test('render', () => {
    const data = [
      {
        title: 'Username',
        key: 'username',
      },
    ];

    const dataSource = {
      results: [
        {
          email: 'bobibasari@gmail.com',
          gender: 'male',
          name: { title: 'Mr', first: 'Bobi', last: 'Basari' },
          registered: { date: '' },
        },
      ],
    };

    const shallow = new ShallowRenderer();
    const tree = shallow.render(
      <Table column={data} dataSource={dataSource} />,
    );
    expect(tree).toMatchSnapshot();
  });

  test('function handleSortAsc', () => {
    const result = Table({ column: [], dataSource: [] });
    result.props.children.props.sortByAsc('name');
    expect(setSortBy).toHaveBeenCalledWith('name');
  });

  test('function handleSortDesc', () => {
    const result = Table({ column: [], dataSource: [] });
    result.props.children.props.sortByDesc('name');
    expect(setSortBy).toHaveBeenCalledWith('name');
  });

  test('render Tables', () => {
    const data = [
      {
        title: 'name',
        ascDesc: true,
        key: 'name',
        data: [{ text: 'Bobi', key: 'name' }],
      },
      {
        title: 'username',
        ascDesc: false,
        key: 'username',
        data: [{ text: 'Bobi', key: 'username' }],
      },
    ];
    Tables.defaultProps.sortByAsc();
    Tables.defaultProps.sortByDesc();

    const shallow = new ShallowRenderer();
    const tree = shallow.render(
      <Tables
        data={data}
        sortBy='name'
        sortByAsc={jest.fn()}
        sortByDesc={jest.fn()}
        sortFormat='asc'
      />,
    );
    expect(tree).toMatchSnapshot();
  });

  test('function sortByAsc', () => {
    const data = [
      {
        title: 'name',
        ascDesc: true,
        key: 'name',
        data: [{ text: 'Bobi', key: 'name' }],
      },
    ];
    const sortByAsc = jest.fn();
    const sortByDesc = jest.fn();

    const result = Tables({
      data,
      sortBy: 'name',
      sortByAsc,
      sortByDesc,
      sortFormat: 'desc',
    });
    result.props.children[0].props.children[0].props.children[1].props.children[0].props.onClick();
    expect(sortByAsc).toHaveBeenCalledWith('name');

    result.props.children[0].props.children[0].props.children[1].props.children[1].props.onClick();
    expect(sortByDesc).toHaveBeenCalledWith('name');
  });
});
