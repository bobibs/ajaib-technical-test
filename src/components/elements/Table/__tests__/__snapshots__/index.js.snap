// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/components/elements/Table render 1`] = `
<div
  className="root"
>
  <Tables
    data={
      Array [
        Object {
          "data": Array [
            Object {
              "key": "username",
              "text": "mrbobi",
            },
          ],
          "key": "username",
          "title": "Username",
        },
      ]
    }
    sortBy=""
    sortByAsc={[Function]}
    sortByDesc={[Function]}
    sortFormat=""
  />
</div>
`;

exports[`src/components/elements/Table render Tables 1`] = `
<div
  className="table"
>
  <div>
    <div
      className="tableHead"
    >
      <div>
        name
      </div>
      <div
        className="tableHeadAction"
      >
        <ImageBox
          className="tableHeadIcon ascIcon tableHeadIconActive"
          format="svg"
          onClick={[Function]}
          src="caret-up"
        />
        <ImageBox
          className="tableHeadIcon descIcon "
          format="svg"
          onClick={[Function]}
          src="caret-down"
        />
      </div>
    </div>
    <div
      className="tableBody"
    >
      <div
        className="tableBodyText"
      >
        <div>
          Bobi
        </div>
      </div>
    </div>
  </div>
  <div>
    <div
      className="tableHead"
    >
      <div>
        username
      </div>
      
    </div>
    <div
      className="tableBody"
    >
      <div
        className="tableBodyText"
      >
        <div>
          Bobi
        </div>
      </div>
    </div>
  </div>
</div>
`;
